File: 21551-naveen-siva-kumar/week7/82-remove-duplicates-from-sorted-list-ii/82-remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        
        ListNode* sentinel = new ListNode(0,head);
        ListNode* first = sentinel;
        
        while(head!=nullptr){
            if(head->next!=nullptr && head->val == head->next->val){
                while(head->next!=nullptr && head->val == head->next->val){
                    head = head->next;
                }
                first->next = head->next;
            }
            else{
                first=first->next;
            }
            head= head->next;
        }
        return sentinel->next;
    }
};

File: 21553 - Sai Nandan D/Week 7/82_Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        pre = dummy = ListNode(0, head)
        while head and head.next:
            if head.val != head.next.val:
                head = head.next
                pre = pre.next
            else:
                while head and head.next and head.val == head.next.val:
                    head = head.next
                pre.next = head.next
                head = head.next
                
        return dummy.next
